File: tools/cdn/sign-url.js
#!/usr/bin/env node
// Minimal Cloud CDN Signed URL generator (Signed URL key on Backend Bucket)
// Usage:
//   node sign-url.js --baseUrl https://cdn.example.com --path /hls/vid/master.m3u8 \
//     --keyName mychannel-key --key <BASE64URL_KEY> [--expiresSec 3600]

const crypto = require('crypto')

function parseArgs() {
  const args = {}
  for (let i = 2; i < process.argv.length; i += 2) {
    const k = process.argv[i]
    const v = process.argv[i + 1]
    if (!k || !v || !k.startsWith('--')) continue
    args[k.slice(2)] = v
  }
  return args
}

function b64urlDecode(input) {
  // convert base64url -> base64
  let b64 = input.replace(/-/g, '+').replace(/_/g, '/')
  while (b64.length % 4) b64 += '='
  return Buffer.from(b64, 'base64')
}

function b64urlEncode(buf) {
  return Buffer.from(buf)
    .toString('base64')
    .replace(/\+/g, '-')
    .replace(/\//g, '_')
    .replace(/=+$/, '')
}

function sign({ baseUrl, path, keyName, key, expiresSec }) {
  if (!baseUrl || !path || !keyName || !key) {
    throw new Error('Missing required args: baseUrl, path, keyName, key')
  }
  const now = Math.floor(Date.now() / 1000)
  const exp = now + (Number(expiresSec || 3600))
  // StringToSign is path + Expires
  const stringToSign = `${path}${exp}`
  const hmacKey = b64urlDecode(key)
  const sig = crypto.createHmac('sha1', hmacKey).update(stringToSign).digest()
  const signature = b64urlEncode(sig)
  const url = new URL(baseUrl)
  url.pathname = path
  url.searchParams.set('Expires', String(exp))
  url.searchParams.set('KeyName', keyName)
  url.searchParams.set('Signature', signature)
  return { url: url.toString(), exp, signature }
}

if (require.main === module) {
  const args = parseArgs()
  const result = sign({
    baseUrl: args.baseUrl,
    path: args.path,
    keyName: args.keyName,
    key: args.key,
    expiresSec: args.expiresSec,
  })
  console.log(result.url)
}

module.exports = { sign }

File: tools/cdn/sign_url.py
#!/usr/bin/env python3
"""Cloud CDN Signed URL generator.

Usage:
  python sign_url.py --base https://cdn.example.com --path /hls/vid/master.m3u8 \
    --key-name mychannel-key --key BASE64URL_KEY --expires 3600
"""

import argparse
import base64
import hashlib
import hmac
import time
from urllib.parse import urlencode, urlunparse


def b64url_decode(s: str) -> bytes:
  s = s.replace('-', '+').replace('_', '/')
  s += '=' * ((4 - len(s) % 4) % 4)
  return base64.b64decode(s)


def b64url_encode(b: bytes) -> str:
  return base64.urlsafe_b64encode(b).decode().rstrip('=')


def sign(base: str, path: str, key_name: str, key: str, expires: int = 3600) -> str:
  now = int(time.time())
  exp = now + int(expires)
  string_to_sign = f"{path}{exp}".encode()
  hkey = b64url_decode(key)
  sig = hmac.new(hkey, string_to_sign, hashlib.sha1).digest()
  signature = b64url_encode(sig)

  query = urlencode({
      'Expires': str(exp),
      'KeyName': key_name,
      'Signature': signature,
  })
  return f"{base}{path}?{query}"


def main():
  ap = argparse.ArgumentParser()
  ap.add_argument('--base', required=True)
  ap.add_argument('--path', required=True)
  ap.add_argument('--key-name', required=True)
  ap.add_argument('--key', required=True, help='base64url (unpadded) key')
  ap.add_argument('--expires', type=int, default=3600)
  args = ap.parse_args()
  print(sign(args.base, args.path, args.key_name, args.key, args.expires))


if __name__ == '__main__':
  main()




